receiver/snmpreceiver/scraper.go

```diff
@@ -39,13 +64,345 @@ func newScraper(logger *zap.Logger, cfg *Config, settings component.ReceiverCrea
+// Keys will be determined from the related attribute config and values will come a few
+// different places.
+// Enum attribute value - comes from the metric config's attribute data
+// Indexed prefix attribute value - comes from the current SNMP data's index and the attribute
+// config's prefix value
+// Indexed OID attribute value - comes from the previously collected indexed attribute data
+// using the current index and attribute config to access the correct value
+func getIndexedDataPointAttributes(
+	configHelper *configHelper,
+	parentOID string,
+	indexString string,
+	indexedAttributeValues map[indexedAttributeKey]string,
+) (map[string]string, error) {
+	datapointAttributes := map[string]string{}
+
+	for _, attribute := range configHelper.getMetricConfigAttributes(parentOID) {
+		attributeName := attribute.Name
+
+		var attributeValue string
+		prefix := configHelper.getAttributeConfigIndexedValuePrefix(attributeName)
+		oid := configHelper.getAttributeConfigOID(attributeName)
+		switch {
+		case prefix != "":
+			attributeValue = prefix + indexString
+		case oid != "":
+			indexedAttrKey := indexedAttributeKey{
+				parentOID: oid,
+				oidIndex:  indexString,
+			}
+			attributeValue = indexedAttributeValues[indexedAttrKey]
+		default:
+			attributeValue = attribute.Value
+		}
+
+		// If no good attribute value could be found
+		if attributeValue == "" {
+			return nil, errors.New(errMsgAttributeEmptyValue)
+		}
+
+		attributeKey := attributeName
+		if value := configHelper.getAttributeConfigValue(attributeKey); value != "" {
+			attributeKey = value
+		}
+		datapointAttributes[attributeKey] = attributeValue
+	}
+
+	return datapointAttributes, nil
+}
+
+// getResourceAttributes creates a map of key/values for all related resource attributes. Keys
+// will come directly from the metric config's resource attribute values. Values will come
+// from the related attribute config's prefix value plus the index OR the previously collected
+// resource attribute indexed data.
+func getResourceAttributes(
+	configHelper *configHelper,
+	parentOID string,
+	indexString string,
+	indexedResourceAttributeValues map[indexedAttributeKey]string,
+) (map[string]string, error) {
+	resourceAttributes := map[string]string{}
+
+	for _, attributeName := range configHelper.getResourceAttributeNames(parentOID) {
+		var attributeValue string
+		prefix := configHelper.getResourceAttributeConfigIndexedValuePrefix(attributeName)
+		oid := configHelper.getResourceAttributeConfigOID(attributeName)
+		switch {
+		case prefix != "":
+			attributeValue = prefix + indexString
+		case oid != "":
+			indexedAttrKey := indexedAttributeKey{
+				parentOID: oid,
+				oidIndex:  indexString,
+			}
+			attributeValue = indexedResourceAttributeValues[indexedAttrKey]
+		default:
+			attributeValue = ""
+		}
+
+		// If no good attribute value could be found
+		if attributeValue == "" {
+			return nil, errors.New(errMsgResourceAttributeEmptyValue)
+		}
```

Comment:
```suggestion
		default:
			return nil, errors.New(errMsgResourceAttributeEmptyValue)
		}
```

The other case where `oid != ""` can also theoretically result in an empty string, but the error messages should be differentiated if that is a concern.

Reply:
@djaglowski i actually fell into this trap while refactoring this to the switch. luckily my unit tests tipped me off something was wrong. when a non empty string OID, we could still end up with a empty attributeValue if something went wrong with that OID. so that's why we can't only bake the error into the switch and have to check after the fact.

Reply:
@djaglowski i mean i can do that. i feel like this is a little "tomato" "tomahto" to me though. they both have their pros and cons and really don't make much of a difference. generally speaking i definitely appreciate your attention to detail in PRs as you have a good eye for making sure things are clean, not redundant, and separated by area of concern. but this sort of complaint is edging a little too close to "PRs which demand change to focus on style preference" to me. (i'll still change this, but do not plan on using a unique error message for the special oid case...it isn't necessary)

Reply:
> @djaglowski i mean i can do that. i feel like this is a little "tomato" "tomahto" to me though. they both have their pros and cons and really don't make much of a difference. generally speaking i definitely appreciate your attention to detail in PRs as you have a good eye for making sure things are clean, not redundant, and separated by area of concern. but this sort of complaint is edging a little too close to "PRs which demand change to focus on style preference" to me. (i'll still change this, but do not plan on using a unique error message for the special oid case...it isn't necessary)

It sounds like it would be helpful to clarify a couple things here. 

This suggestion was a relatively small one, but it was not arbitrary. Rather, I found the code difficult to understand because we were assigning a value _before_ checking it to see if it represented an error case, Not only that, but the error check was happening elsewhere and grouped with an error check for a different failure mode. In general, we should strive for clarity in our code. Maybe that could be described as a style preference, but there is a meaningful difference.

I also want to be clear that I am not complaining about your code. My purpose in asking for changes to is help make the code more maintainable. This component is particularly is more complex than most, so it is especially important that it be clear. You may have understood it perfectly well since the first draft, but the business logic has only started to become clear to me after many iterations of improving the code. Regardless of the number of iterations that we have gone through, I will ask for clarification and/or changes when I am confused by code for which I am assuming responsibility.